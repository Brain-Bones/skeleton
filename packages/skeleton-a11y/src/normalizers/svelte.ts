import { createNormalizer } from './create.js';

const normalizeSvelteAttributes = createNormalizer({});

export { normalizeSvelteAttributes };
