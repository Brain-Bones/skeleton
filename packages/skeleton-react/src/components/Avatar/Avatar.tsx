"use client";

import React from "react";
import { AvatarProps } from "./types";

export const Avatar: React.FC<AvatarProps> = ({
    // Root
    base = 'overflow-hidden isolate',
    background = 'bg-surface-400-600',
    size = 'size-16',
    font = '',
    border = '',
    rounded = 'rounded-full',
    shadow = '',
    classes = '',
    // Image
    imageSrc = '',
    imageAlt = '',
    imageFilter = '',
    imageBase = 'w-full object-cover',
    imageClasses = '',
    // Children
    children
}): React.ReactElement => {
    return (
        <figure className={`${base} ${background} ${size} ${font} ${border} ${rounded} ${shadow} ${classes}`} data-testId="avatar">
            {imageSrc ? (
                <img className={`${imageBase} ${imageClasses}`} src={imageSrc} alt={imageAlt} style={{ filter: imageFilter ? `url(${imageFilter})` : undefined }} />
            ) : (
                children
            )}
	    </figure>  
    );
};