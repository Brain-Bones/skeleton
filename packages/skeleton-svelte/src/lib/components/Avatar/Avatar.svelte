<script lang="ts">
  import type { AvatarProps } from './types.js';

  let {
    src = '',
    alt = '',
    filter = '',
    // Root
    base = 'overflow-hidden isolate',
    background = 'bg-surface-400-600',
    size = 'size-16',
    font = '',
    border = '',
    rounded = 'rounded-full',
    shadow = '',
    classes = '',
    // Image
    imageBase = 'w-full object-cover',
    imageClasses = '',
    // Snippets
    children
  }: AvatarProps = $props();
</script>

<!-- @component An image with a fallback for representing the user. -->

<figure class="{base} {background} {size} {font} {border} {rounded} {shadow} {classes}" data-testId="avatar">
  {#if src}
    <img class="{imageBase} {imageClasses}" {src} {alt} style:filter={filter && `url(${filter})`} />
  {:else if children}
    {@render children()}
  {/if}
</figure>
