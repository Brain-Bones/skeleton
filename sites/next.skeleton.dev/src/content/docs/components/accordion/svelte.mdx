---
layout: '@layouts/LayoutDoc.astro'
---

export const components = componentSet;

import Example from '@examples/components/accordions/Example.svelte';
import ExampleRaw from '@examples/components/accordions/Example.svelte?raw';

{/* --- */}

<Preview client:load>
	<Fragment slot="preview">
		<Example client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleRaw} lang="svelte" />
	</Fragment>
</Preview>

## Open State

Each Item `id` added to the Accordion `value` array will be open by default.

### State Rune

```ts
const value = $state(['club']);
```
```tsx
<Accordion value={value}>...</Accordion>
```

### Harcoded

```tsx
<Accordion value={['club']}>...</Accordion>
```

## Multiple Items

Use `multiple` to enable multiple items to be opened at once.

```html
<Accordion multiple>...</Accordion>
```

## Custom Icons

```svelte
<script lang="ts">
	import Plus from 'lucide-svelte/icons/plus';
	import Minus from 'lucide-svelte/icons/minus';
	import Club from 'lucide-svelte/icons/club';
</script>

<Accordion>
    {#snippet iconOpen()}<Plus size={16} />{/snippet}
    {#snippet iconClosed()}<Minus size={16} />{/snippet}
    <!-- ... -->
	<AccordionItem id="club">
    	{#snippet controlLead()}<Club size={24} />{/snippet}
    	<!-- ... -->
	<AccordionItem>
</Accordion>
```
