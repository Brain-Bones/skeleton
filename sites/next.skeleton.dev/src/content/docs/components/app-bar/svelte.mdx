---
layout: '@layouts/LayoutDoc.astro'
---

export const components = componentSet;

import Example from '@examples/components/app-bars/Example.svelte';
import ExampleRaw from '@examples/components/app-bars/Example.svelte?raw';
import ExampleToolbar from '@examples/components/app-bars/ExampleToolbar.svelte';
import ExampleToolbarRaw from '@examples/components/app-bars/ExampleToolbar.svelte?raw';
import ExampleResponsive from '@examples/components/app-bars/ExampleResponsive.svelte';
import ExampleResponsiveRaw from '@examples/components/app-bars/ExampleResponsive.svelte?raw';
import AppBarDocs from '@component-docs/svelte/AppBar.json';

<Preview client:load>
	<Fragment slot="preview">
		<Example client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleRaw} lang="svelte" />
	</Fragment>
</Preview>

## Toolbar

<Preview client:load>
	<Fragment slot="preview">
		<ExampleToolbar client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleToolbarRaw} lang="svelte" />
	</Fragment>
</Preview>

## Responsive

Use Tailwind's [responsive design](https://tailwindcss.com/docs/responsive-design) breakpoints to adjust for various screen sizes.

<Preview client:load>
	<Fragment slot="preview">
		<ExampleResponsive client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleResponsiveRaw} lang="svelte" />
	</Fragment>
</Preview>

## API Reference

<ApiTable
	types={['AppBarProps', 'ToolBarProps', 'ToolbarLeadProps', 'ToolbarCenterProps', 'ToolbarTrailProps', 'AppBarHeadlineProps']}
	schema={AppBarDocs}
/>
