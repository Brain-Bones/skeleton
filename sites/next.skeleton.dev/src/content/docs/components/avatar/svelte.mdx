---
layout: '@layouts/LayoutDoc.astro'
---

export const components = componentSet;

import Example from '@examples/components/avatars/Example.svelte';
import ExampleRaw from '@examples/components/avatars/Example.svelte?raw';

<Preview client:load>
  <Fragment slot="preview">
    <Example client:load />
  </Fragment>
  <Fragment slot="code">
    <Code code={ExampleRaw} lang="svelte" />
  </Fragment>
</Preview>

## Fallback

When the `src` prop is _undefined_, default children will show. This can be used for initials, icons, or fallback images.

```html
<Avatar>SK</Avatar>
```

```html
<Avatar>ðŸ’€</Avatar>
```
