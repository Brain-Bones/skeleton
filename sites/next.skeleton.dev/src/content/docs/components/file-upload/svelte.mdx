---
layout: '@layouts/LayoutDoc.astro'
---

export const components = componentSet;

import Example from '@examples/components/file-upload/Example.svelte';
import ExampleRaw from '@examples/components/file-upload/Example.svelte?raw';
import ExampleFull from '@examples/components/file-upload/ExampleFull.svelte';
import ExampleFullRaw from '@examples/components/file-upload/ExampleFull.svelte?raw';
import ExampleButton from '@examples/components/file-upload/ExampleButton.svelte';
import ExampleButtonRaw from '@examples/components/file-upload/ExampleButton.svelte?raw';
import ExampleDisabled from '@examples/components/file-upload/ExampleDisabled.svelte';
import ExampleDisabledRaw from '@examples/components/file-upload/ExampleDisabled.svelte?raw';
import ExampleRtl from '@examples/components/file-upload/ExampleRtl.svelte';
import ExampleRtlRaw from '@examples/components/file-upload/ExampleRtl.svelte?raw';

<Preview client:load>
	<Fragment slot="preview">
		<Example client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleRaw} lang="svelte" />
	</Fragment>
</Preview>

<Preview client:load>
	<Fragment slot="preview">
		<ExampleFull client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleFullRaw} lang="svelte" />
	</Fragment>
</Preview>

## Button

Use the default `children` slot to overwrite the default dropzone interface. Allow for a custom interface.

<Preview client:load>
	<Fragment slot="preview">
		<ExampleButton client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleButtonRaw} lang="svelte" />
	</Fragment>
</Preview>

## Disabled

<Preview client:load>
	<Fragment slot="preview">
		<ExampleDisabled client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleDisabledRaw} lang="svelte" />
	</Fragment>
</Preview>

## RTL

<Preview client:load>
	<Fragment slot="preview">
		<ExampleRtl client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleRtlRaw} lang="svelte" />
	</Fragment>
</Preview>
