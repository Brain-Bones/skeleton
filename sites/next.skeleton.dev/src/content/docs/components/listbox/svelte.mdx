---
layout: '@layouts/LayoutDoc.astro'
---

export const components = componentSet;

import ExampleSingle from '@examples/components/listbox/ExampleSingle.svelte';
import ExampleSingleRaw from '@examples/components/listbox/ExampleSingle.svelte?raw';

import ExampleMultiple from '@examples/components/listbox/ExampleMultiple.svelte';
import ExampleMultipleRaw from '@examples/components/listbox/ExampleMultiple.svelte?raw';

{/* --- */}

## Single

<Preview client:load>
	<Fragment slot="preview">
		<ExampleSingle client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleSingleRaw} lang="svelte" />
	</Fragment>
</Preview>

## Multiple

<Preview client:load>
	<Fragment slot="preview">
		<ExampleMultiple client:load />
	</Fragment>
	<Fragment slot="code">
		<Code code={ExampleMultipleRaw} lang="svelte" />
	</Fragment>
</Preview>