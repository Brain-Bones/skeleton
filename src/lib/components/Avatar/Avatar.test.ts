import { render, screen, fireEvent } from '@testing-library/svelte';
import { describe, it, expect, vitest, afterEach } from 'vitest';

import Avatar from '$lib/components/Avatar/Avatar.svelte';
import { filter } from '$lib/actions/Filters/filter';

const imgTextSrc = 'https://i.pravatar.cc/512?img=48';

const cBase = 'flex aspect-square text-surface-50 font-semibold justify-center items-center overflow-hidden isolate';
const cImage = 'w-full h-full object-cover';

const props = {
	background: 'bg-surface-500',
	width: 'w-12',
	border: '',
	rounded: 'rounded-full',
	shadow: 'shadow-xl',
	cursor: 'cursor-pointer',
	class: 'my-custom-class'
};

const defaultProps = {
	background: 'bg-surface-400-500-token',
	width: 'w-12',
	border: '',
	rounded: 'rounded-full',
	shadow: '',
	cursor: ''
};

const imageProps = {
	...props,
	// eslint-disable-next-line @typescript-eslint/no-empty-function
	action: filter,
	actionParams: '#Apollo',
	src: imgTextSrc,
	alt: 'Avatar image',
	style: 'width: 40px; height: 40px; border: 2px solid #b6bcd1;'
};

const initialsProps = { ...props, fill: 'fill-slate-50', initials: 'LA' };

const getWrapper = () => screen.getByTestId('avatar');

const getInitialsSvg = (text: string) => ({
	initialsSvg: screen.getByTestId('avatar-initials'),
	svgText: screen.getByText(text)
});

//TODO: Consider moving into global utility functions, and add test cases to it
const textToInitials = (text: string) => String(text).substring(0, 2).toUpperCase();
/**
 * @description Regex pattern to match a string in a text no matter the location of it
 * @param stringToMatch
 * @returns RegExp
 */
const pattern = (stringToMatch: string) => new RegExp('\\b' + stringToMatch + '\\b');

describe('Avatar.svelte', () => {
	afterEach(() => {
		vitest.clearAllMocks();
	});

	/**  ------------------------ Testing Overall Component Starts ------------------------ */
	it('Renders with minimal props', async () => {
		//Arrange
		render(Avatar);

		//Act
		const [wrapper, { initialsSvg, svgText }] = [getWrapper(), getInitialsSvg('AB')];

		//Assert
		expect(wrapper.className).toMatch(cBase);
		expect(wrapper.className).toMatch(pattern('avatar'));
		expect(svgText.textContent).toStrictEqual('AB');
		expect(initialsSvg.classList.value).toMatch(pattern('avatar-initials w-full h-full'));
		Object.values(defaultProps).forEach((className) => expect(wrapper.className).toMatch(pattern(className)));
	});

	it('Renders component with all props', async () => {
		//Arrange
		render(Avatar, { ...initialsProps, ...imageProps });

		//Act
		const wrapper = getWrapper();
		const image = screen.getByAltText(imageProps.alt);

		//Assert
		expect(wrapper.className).toMatch(pattern(cBase));
		expect(wrapper.className).toMatch(pattern('avatar'));
		expect(image).toHaveProperty('src', imageProps.src);
		expect(image).toHaveProperty('alt', imageProps.alt);
		Object.values(props).forEach((className) => expect(wrapper.className).toMatch(pattern(className)));
		//TODO: Consider adding jest-dom to enable many more functionalities like testing styles as such:
		//expect(image).toHaveStyle(imageProps.style)
	});

	it('Renders img when both img and initials are provided', async () => {
		//Arrange
		render(Avatar, { ...initialsProps, ...imageProps });

		//Act
		const image = screen.queryByAltText(imageProps.alt);
		const initialsSvg = screen.queryByTestId('avatar-initials');

		//Assert
		expect(image).toBeTruthy();
		expect(initialsSvg).toBeFalsy();
	});

	it('Renders wrapper with custom class', async () => {
		//Arrange
		const myClass = 'try-this-now';
		render(Avatar, { ...props, class: myClass });

		//Act
		const wrapper = getWrapper();

		//Assert
		expect(wrapper.className).toMatch(pattern(myClass));
	});

	//TODO: Consider adding userEvent from testing-library that makes testing events much more practical (Declarative)
	it('Fires events properly', async () => {
		//Arrange
		const [onClick, onKeyDown, onKeyUp, onKeyPress] = [vitest.fn(), vitest.fn(), vitest.fn(), vitest.fn()];
		const { component } = render(Avatar, { ...imageProps });
		component.$on('click', onClick);
		component.$on('keydown', onKeyDown);
		component.$on('keyup', onKeyUp);
		component.$on('keypress', onKeyPress);

		//Act
		const wrapper = getWrapper();

		//Act
		fireEvent.click(wrapper);
		//Assert
		expect(onClick).toHaveBeenCalledOnce();

		//Act
		fireEvent.keyDown(wrapper);
		//Assert
		expect(onKeyDown).toHaveBeenCalledOnce();

		//Act
		fireEvent.keyUp(wrapper);
		//Assert
		expect(onKeyUp).toHaveBeenCalledOnce();

		//Act
		fireEvent.keyPress(wrapper);
		//Assert
		expect(onKeyPress).toHaveBeenCalledOnce();
	});
	/**  ------------------------ Testing Overall Component Ends ------------------------ */

	/**  ------------------------ Testing Initials Starts ------------------------ */

	it('Renders initials with all props', async () => {
		//Arrange
		render(Avatar, initialsProps);

		//Act
		const { initialsSvg, svgText } = getInitialsSvg(initialsProps.initials);

		//Assert
		expect(svgText.textContent).toStrictEqual(initialsProps.initials);

		expect(svgText.classList.value).toMatch(pattern('avatar-text'));
		expect(svgText.classList.value).toMatch(pattern(initialsProps.fill));
		expect(initialsSvg.classList.value).toMatch(pattern('avatar-initials w-full h-full'));
	});

	it('Renders initials', async () => {
		//Arrange
		const initials = 'SD';
		render(Avatar, { initials });

		//Act
		const { svgText } = getInitialsSvg(initials);

		//Assert
		expect(svgText.textContent).toStrictEqual(initials);
	});

	it('Renders initials capitalized', async () => {
		//Arrange
		const initials = 'sd';
		render(Avatar, { initials });

		//Act
		const { svgText } = getInitialsSvg(initials.toUpperCase());

		//Assert
		expect(svgText.textContent).toStrictEqual(initials.toUpperCase());
	});

	it('Renders any string into initials', async () => {
		//Arrange
		const text = 'Hello';
		const initials = textToInitials(text);
		render(Avatar, { initials: text });

		//Act
		const { svgText } = getInitialsSvg(initials);

		//Assert
		expect(svgText.textContent).toStrictEqual(initials);
	});

	/**  ------------------------  Testing Initials Ends ------------------------ */

	/**  ------------------------  Testing  {Image} Starts ------------------------ */

	//It renders custom action and ActionParam

	it('Renders image', async () => {
		//Arrange
		render(Avatar, { ...imageProps });

		//Act
		const img = screen.getByAltText(imageProps.alt);

		//Assert
		expect(img).toHaveProperty('src', imageProps.src);
		expect(img).toHaveProperty('alt', imageProps.alt);
		expect(img.className).toMatch(pattern('avatar-image'));
		expect(img.className).toMatch(pattern(cImage));
	});

	it('Renders with both action and actionParam from Skeleton', async () => {
		//Arrange
		const mockedFilterFn = vitest.fn().mockImplementation(filter);
		render(Avatar, { ...imageProps, action: mockedFilterFn });

		//Act
		const img = screen.getByAltText(imageProps.alt);

		//Assert
		expect(mockedFilterFn).toHaveBeenCalled();
		expect(mockedFilterFn).toHaveBeenCalledWith(img, imageProps.actionParams);

		//TODO: Missing styles test(s), pending jest-dom
	});

	it('Renders custom action and actionParam', async () => {
		//Arrange
		const mockedFilterFn = vitest.fn();
		const customParams = { action: mockedFilterFn, actionParams: 'HIIII' };
		render(Avatar, { ...imageProps, ...customParams });

		//Act
		const img = screen.getByAltText(imageProps.alt);

		//Assert
		expect(mockedFilterFn).toHaveBeenCalled();
		expect(mockedFilterFn).toHaveBeenCalledWith(img, customParams.actionParams);
		//TODO: Missing styles test(s), pending jest-dom
	});
	//TODO: Missing test(s)
	//It renders image with custom styles, pending jest-dom

	/**  ------------------------ Testing Image Ends ------------------------ */
});
