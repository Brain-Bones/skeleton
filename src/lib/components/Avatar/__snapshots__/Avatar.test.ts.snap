// Vitest Snapshot v1

exports[`Avatar.svelte > Renders component with all props > Component, with all props 1`] = `
<figure
  class="avatar flex aspect-square text-surface-50 font-semibold justify-center items-center overflow-hidden isolate bg-surface-500 w-12  rounded-full shadow-xl cursor-pointer my-custom-class"
>
  <img
    alt="Avatar image"
    class="avatar-image w-full h-full object-cover"
    src="https://i.pravatar.cc/512?img=48"
    style="filter: url(\\"#Apollo\\")"
  />
</figure>
`;

exports[`Avatar.svelte > Renders image > Img, with all props 1`] = `
<img
  alt="Avatar image"
  class="avatar-image w-full h-full object-cover"
  src="https://i.pravatar.cc/512?img=48"
  style="filter: url(\\"#Apollo\\")"
/>
`;

exports[`Avatar.svelte > Renders initials with all props > Svg, with all props 1`] = `
<svg
  class="avatar-initials w-full h-full"
  viewBox="0 0 512 512"
>
  <text
    class="avatar-text fill-slate-50"
    dominant-baseline="middle"
    font-size="150"
    font-weight="bold"
    text-anchor="middle"
    x="50%"
    y="50%"
  >
    LA
  </text>
</svg>
`;
